cpp/src/io/parquet/page_data.cu

```diff
@@ -1845,8 +1940,691 @@ __global__ void __launch_bounds__(block_size)
+  // for all n
+  __shared__ __align__(8) int64_t prefix_lens[warp_size];
+  __shared__ __align__(8) uint8_t const* offsets[warp_size];
+
+  uint32_t const value_count = dba->prefixes.value_count;
+  uint32_t const ln_idx      = start_idx + lane_id;
+  uint32_t const src_idx     = rolling_index(ln_idx);
+  uint64_t prefix_len        = ln_idx < value_count ? dba->prefixes.value[src_idx] : 0;
+  uint8_t* const lane_out    = ln_idx < value_count ? strings_out + dba->offset[src_idx] : nullptr;
+
+  prefix_lens[lane_id] = prefix_len;
+  offsets[lane_id]     = lane_out;
+
+  // if all prefix_len's are zero, then there's nothing to do
+  if (__all_sync(0xffff'ffff, prefix_len == 0)) { return; }
+
+  // find a neighbor to the left that has a prefix length less than this lane. once that
+  // neighbor is complete, this lane can be completed.
+  int blocker = lane_id - 1;
+  while (blocker > 0 && prefix_lens[blocker] != 0 && prefix_len <= prefix_lens[blocker]) {
+    blocker--;
+  }
+
+  // fill in lane 0 (if necessary)
+  if (lane_id == 0 && prefix_len > 0) {
+    memcpy(lane_out, last_string, prefix_len);
+    prefix_lens[0] = prefix_len = 0;
+  }
+  __syncwarp();
+
+  // now fill in blockers until done
+  for (uint32_t i = 1; i < warp_size && i + start_idx < value_count; i++) {
+    if (prefix_len != 0 && prefix_lens[blocker] == 0 && lane_out != nullptr) {
+      memcpy(lane_out, offsets[blocker], prefix_len);
+      prefix_lens[lane_id] = prefix_len = 0;
+    }
+
+    // check for finished
+    if (__all_sync(0xffff'ffff, prefix_len == 0)) { return; }
+  }
+}
+
+// given prefix and suffix state, plus pointers to the page suffix data
+// and page output data, calculate a batch of output strings.  called
+// by all threads in a warp.
+// updates suffix_data and last_string pointers.
+__device__ void CalculateStringValues(delta_byte_array_state_s* dba,
+                                      uint8_t const*& suffix_data,
+                                      uint8_t* strings_out,
+                                      uint8_t const*& last_string,
+                                      uint32_t start_idx,
+                                      uint32_t lane_id)
+{
+  using cudf::detail::warp_size;
+  __shared__ __align__(8) uint64_t strings_offset;
+
+  auto* const prefix_db = &dba->prefixes;
+  auto* const suffix_db = &dba->suffixes;
+
+  if (start_idx >= suffix_db->value_count) { return; }
+  uint32_t const end_idx = start_idx + suffix_db->values_per_mb;
+
+  for (int idx = start_idx; idx < end_idx; idx += warp_size) {
+    uint32_t const ln_idx = idx + lane_id;
+
+    // calculate offsets into suffix data
+    uint32_t const src_idx    = rolling_index(ln_idx);
+    uint64_t const suffix_len = ln_idx < suffix_db->value_count ? suffix_db->value[src_idx] : 0;
+    uint64_t suffix_off       = 0;
+    __shared__ cub::WarpScan<uint64_t>::TempStorage temp_storage;
+    cub::WarpScan<uint64_t>(temp_storage).ExclusiveSum(suffix_len, suffix_off);
```

Comment:
It looks like the only ways into this function involve one warp, so this will work. But it does seem vulnerable to someone rearranging things later on where multiple warps are used.

Reply:
There's a comment at the top that the function is only supported on a single warp. Not sure what else can be done to guard it.  No more vulnerable than `gpuDecodeLevels`...I guess I can make the warning stand out more in the docstring (well, once there is a docstring).